src/UKHO.ERPFacade.API/Helpers/SapMessageBuilder.cs

```diff
@@ -0,0 +1,233 @@
+            }
+
+            XmlNode xmlNode = SortXmlPayload(actionItemNode);
+
+            XmlNode noOfActions = soapXml.SelectSingleNode(XpathNoOfActions);
+            XmlNode corrId = soapXml.SelectSingleNode(XpathCorrId);
+            XmlNode recDate = soapXml.SelectSingleNode(XpathRecDate);
+            XmlNode recTime = soapXml.SelectSingleNode(XpathRecTime);
+
+            corrId.InnerText = traceId;
+            noOfActions.InnerText = xmlNode.ChildNodes.Count.ToString();
+            recDate.InnerText = DateTime.UtcNow.ToString("yyyyMMdd");
+            recTime.InnerText = DateTime.UtcNow.ToString("hhmmss");
+
+            IM_MATINFONode.AppendChild(xmlNode);
+
+            return soapXml;
+        }
+
+        private static XmlNode SortXmlPayload(XmlNode actionItemNode)
+        {
+            List<XmlNode> actionItemList = new();
+            int sequenceNumber = 1;
+
+            foreach (XmlNode subNode in actionItemNode)
+            {
+                actionItemList.Add(subNode);
+            }
+
+            var sortedActionItemList = actionItemList.Cast<XmlNode>().OrderBy(x => Convert.ToInt32(x.SelectSingleNode(ActionNumber).InnerText)).ToList();
+
+            foreach (XmlNode actionItem in sortedActionItemList)
+            {
+                actionItem.SelectSingleNode(ActionNumber).InnerText = sequenceNumber.ToString();
+                sequenceNumber++;
+            }
+
+            foreach (XmlNode actionItem in sortedActionItemList)
+            {
+                actionItemNode.AppendChild(actionItem);
+            }
+            return actionItemNode;
+        }
+
+        private static XmlElement BuildAction(XmlDocument soapXml, Scenario scenario, SapAction action, string cell)
+        {
+            XmlElement itemNode = soapXml.CreateElement(Item);
+
+            XmlElement actionNumberNode = soapXml.CreateElement(ActionNumber);
+            actionNumberNode.InnerText = action.ActionNumber.ToString();
+
+            XmlElement actionNode = soapXml.CreateElement(Action);
+            actionNode.InnerText = action.Action.ToString();
+
+            XmlElement productNode = soapXml.CreateElement(Product);
+            productNode.InnerText = action.Product.ToString();
+
+            itemNode.AppendChild(actionNumberNode);
+            itemNode.AppendChild(actionNode);
+            itemNode.AppendChild(productNode);
+
+            List<(int sortingOrder, XmlElement itemNode)> actionAttributeList = new();
+
+            foreach (var node in action.Attributes.Where(x => x.Section == ProductSection))
+            {
+                XmlElement itemSubNode = soapXml.CreateElement(node.XmlNodeName);
+
+                if (node.IsRequired)
+                {
+                    if (node.XmlNodeName == ReplacedBy)
+                    {
+                        itemSubNode.InnerText = string.IsNullOrWhiteSpace(cell.ToString()) ? string.Empty : cell.ToString().Substring(0, Math.Min(250, cell.ToString().Length));
```

Comment:
Can we make one method to get this value and use it wherever needed as same is repeated couple of times.

Reply:
done.